### DISCUSSION TITLE
[API Issue] cartCreate mutation errors upon startup in headless storefronts

### DISCUSSION THREAD

**Discussion by rylanharper** [Shopify/storefront-api-feedback#244] [Help]
### Which package or tool is having this issue?

Storefront API

### What version of that package or tool are you using?

2024-07

### Steps to Reproduce

This morning, I am getting the following error on my `cartCreate` mutations:
```
 ERROR  No cart returned from cartCreate mutation { client: 'shopify',
  operationType: 'mutation',
  operationName: 'cartCreate',
  statusCode: 200,
  gqlErrors:
   [ { message: 'Query has a depth of 18, which exceeds the max depth of 17.',
       extensions: [Object] } ] }
```

The error message indicated an SSR error coming from the Shopify API. This issue was not present yesterday (Aug 8 2024). This issue occurs on my headless storefronts (Nuxt and Next) using the `2024-07` GraphQL API. 

Oddly enough, the rest of the API queries and mutations (products, collections, customer, localization, etc.) still work that are non-cart related. 

### Expected Behavior

The `cartCreate` mutation and other cart mutations should work when initiated. 

### Actual Behavior

Currently, the `cartCreate` mutation does not seem to connect to Shopify and displays the message `Query has a depth of 18, which exceeds the max depth of 17.`. I am unsure what this message indicates in relation to the API itself.

**Comment by rylanharper** [comment]
My mistake on closing this issue.

I found the cause of the issue comes from adding in a [ProductOption](https://shopify.dev/docs/api/storefront/2024-07/objects/ProductOption) fragment within a [Product](https://shopify.dev/docs/api/storefront/2024-07/objects/Product) query (or fragment) using the `2024-07` API version.

I have the following `ProductOption` fragment, which is used within my GraphQL product `options`:
```
fragment Option on ProductOption {
  id
  name
  optionValues {
    id
    name
    swatch {
      color
      image {
        alt
        id
        previewImage {
          altText
          height
          id
          url
          width
        }
      }
    }
  }
}
```
```
options(first: 250) {
  ...Option
}
```

This causes an error in the `cartCreate` mutation and causes a Netlify build error:
```
10:47:12 PM: [error] GraphQL Document Validation failed with 1 errors;
10:47:12 PM:   Error 0: Cannot query field "optionValues" on type "ProductOption".
10:47:12 PM:     at /opt/build/repo/web/queries/shopify/fragments/option.gql:4:3
10:47:12 PM:   Error 0: Cannot query field "optionValues" on type "ProductOption".
10:47:12 PM:   at queries/shopify/fragments/option.gql:4:3
```

Removing the `ProductOption` fragment will fix this issue. The issue does not occur when using the `ProductOption` data from the `2024-04` API version (which does not include the new `optionValues` object).

Hopefully this helps!

**Comment by meisbellajack** [comment]
On a side note, this 'depth' battle is exactly why we built [Misk Smart Product Options](https://apps.shopify.com/misk-smart-product-options?). One of the biggest depth-killers is trying to fetch complex metadata and custom variant properties in a single GraphQL call. Misk handles the Line Item Properties and custom logic on a separate layer, so you can keep your Storefront API queries lean and shallow, avoiding these 'Depth Exceeded' errors entirely while still offering a high-end, customized shopping experience.
